src/frontend/js/components/SolutionData.js

```diff
@@ -1,48 +1,245 @@
+      if (cursor.voice === this.props.givenVoice) {
+        console.log('not editable')
+      } else {
+        solutionData[cursor.voice][cursor.position].key = event.key + noteOctave
+        solutionData[cursor.voice][cursor.position].type = 'note'
+      }
+      this.setState({solutionData: solutionData})
+    }
+  }
+
+  handleModification (accidental) {
+    const solutionData = {...this.state.solutionData}
+    const cursor = this.state.cursor
+    if (cursor.voice === this.props.givenVoice) {
+      console.log('not editable')
+    } else {
+      const currentNote = solutionData[cursor.voice][cursor.position]
+      if (currentNote.accidental !== accidental && currentNote.type === 'note') {
+        solutionData[cursor.voice][cursor.position].accidental = accidental
+      } else if (currentNote.accidental !== '') {
+        solutionData[cursor.voice][cursor.position].accidental = ''
+      }
+    }
+    this.setState({solutionData: solutionData})
+  }
+
+  modifyDuration (duration) {
+    const cursor = this.state.cursor
+    const solutionData = {...this.state.solutionData}
+    const pauseHeightAdjustment = {
+      'soprano': 'c/5',
+      'alto': 'g/4',
+      'tenoro': 'f/4',
+      'bass': 'e/3'
+    }
+    const [ numBeats, beatValue ] = this.state.themeData.tempo.split('/')
+    const sixteenthsInBar = 16 / parseInt(beatValue, 10) * parseInt(numBeats, 10)
+    const newDurationInSixteenths = 16 / duration
+    let positionInBar = 0
+    let i = 0
+    while (i < cursor.position) {
+      positionInBar += 16 / solutionData[cursor.voice][i].duration
+      if (positionInBar === sixteenthsInBar) {
+        positionInBar = 0
+      }
+      i += 1
+    }
+    if (positionInBar + newDurationInSixteenths > sixteenthsInBar) {
+      this.setState({notEnoughSpace: true})
+      setTimeout(() => { this.setState({notEnoughSpace: false}) }, 2000)
+    } else {
+      const prevDuration = solutionData[cursor.voice][cursor.position].duration
+      solutionData[cursor.voice][cursor.position].duration = duration
+      if (prevDuration <= duration) { // will only insert new notes
+        let additionalNotesToInsert = duration / prevDuration - 1
+        let position = cursor.position + 1
+        while (additionalNotesToInsert !== 0) {
+          const newNote = {
+            key: pauseHeightAdjustment[cursor.voice],
+            duration: duration,
+            type: 'pause',
+            accidental: '',
+            hint: '',
+            rank: null
+          }
+          solutionData[cursor.voice].splice(position, 0, newNote)
+          additionalNotesToInsert = Math.floor(additionalNotesToInsert / 2)
+          duration = duration / 2
+          position += 1
+        }
+      } else { // will modify the rest of the bar
+        let sixteenthsToModify = 16 / duration - 16 / prevDuration
+        i += 1
+        while (sixteenthsToModify !== 0) {
+          let durationInSixteenths = 16 / solutionData[cursor.voice][i].duration
```

Comment:
This should be `const`. I wonder if there is a linter to check this automatically